### PR TITLE
Add equality witness type to the standard library

### PR DESCRIPTION
**Pull request by nojb**
See discussion in #11564. It is argued there that this type is fundamental enough to go in `Stdlib`. Further constructions based on this type would go into their own module (but this can be done in a separate PR).

cc @garrigue @yallop

### CODE REVIEW

**Comment by dbuenzli** [comment]
Personally I still think it's a bad idea to add more the initial scope as a matter of principle, especially if we don't have a good idea of how this:

> Further constructions based on this type would go into their own module (but this can be done in a separate PR).

Is going to look like. I don't think it's horrible to have that name prefixed by a module and OCaml's [own manual](https://v2.ocaml.org/releases/4.14/htmlman/gadts-tutorial.html#s:gadts-advexamples) examples uses `Eq` for this construction which I find much clearer in pattern matching scenarios than `Refl`.

**Comment by garrigue** [comment]
Ah, `Refl` vs. `Eq`.
I had been using `Eq` for a long time, but I suppose I was eventually won over by Coq and Agda: an equality witness is created by using reflexivity, so the name is `Refl`. OCaml can of course choose to go its own way, but I'm not sure we have much to gain there.

As to adding something else to the standard library, I would expect a few more years before we decide on anything.
Basically, you can already do most things by just pattern-matching onto `Refl`, so anything else would have more specific applications in mind.

**Comment by alainfrisch** [comment]
I'm biased towards `Eq` (because that's what we have been using for ages at LexiFi), but I don't find the argument about reflexivity being how equality witnesses are created very convincing.  Those values are witnesses of equality, not witnesses of reflexivity.  If a function dynamically checks for equality and returns an `Eq.t option`, it feels more natural to check whether there is "some equality" than "some reflexivity".

Moreover, I'm not sure that "reflexivity" is a concept universally known to the programming audience (and so, `Refl` might be a bit mysterious), while "equality" is a much more common term.  Btw, `Equal` might be short enough to avoid the `Eq` abbreviation.

**Comment by dbuenzli** [comment]
> As to adding something else to the standard library, I would expect a few more years before we decide on anything.
> Basically, you can already do most things by just pattern-matching onto `Refl`, so anything else would have more specific applications in mind.

That doesn't mean you shouldn't try to plan for it. By your account we could also just say basically you can already do most things by simply defining the constructor yourself so let's not add it. 

This is a type level trick, why not make it obvious in the name ? `Type.Eq`, `Type.Equal` or `Type.Refl` (I care more about the prefix than the actual name of the variant) seems much better to me than an unprefixed variant. See my [comment on the original issue](https://github.com/ocaml/ocaml/issues/11564#issuecomment-1261396406).

**Comment by ferminr** [comment]
> This is a type level trick, why not make it obvious in the name ? `Type.Eq`, `Type.Equal` or `Type.Refl` (I care more about the prefix than the actual name of the variant) seems much better to me than an unprefixed variant. See my [comment on the original issue](https://github.com/ocaml/ocaml/issues/11564#issuecomment-1261396406).

FWIW, Haskell also uses `Refl` as the name of the constructor, defined inside module `Data.Type.Equality`. (https://hackage.haskell.org/package/base-4.17.0.0/docs/src/Data.Type.Equality.html#%3A~%3A)

**Comment by dbuenzli** [comment]
If `Refl` is to be chosen, then I think Haskell's scheme makes sense, `Refl` lacks context, `Type.Refl` applies to the wrong thing, `Type.Equality.Refl` starts to make sense (at least to me). 

So what about:

```ocaml 
module Type : sig 
  module Equality : sig 
    type (_, _) t = Refl : ('a, 'a) t
  end 
end
```

**Comment by alainfrisch** [comment]
> Type.Equality.Refl

I find this one both more verbose and more cryptic than `Type.Equal`.

**Comment by dbuenzli** [comment]
> I find this one both more verbose and more cryptic than `Type.Equal`.

I don't buy the verbosity argument, it's not as if I use this every two lines of code, I don't think a short name is needed. But I'm not attached to `Refl` name, replace it with `Type.Equality.Equal` if you wish.

The good thing with that proposal is that it nudges us to a well organized `Type` module with submodules that respect the `M.t` convention and where [your additional functions](https://github.com/ocaml/ocaml/issues/11564#issuecomment-1257614318) could be easily added (I personally never felt the need for these so I can't comment on their usefullness).

File: stdlib/stdlib.mli [hunk 1]
```diff
@@ -1369,6 +1369,12 @@ val at_exit : (unit -> unit) -> unit
    The functions are called in 'last in, first out' order: the
    function most recently added with [at_exit] is called first. *)
 
+(** {1 Type equality witness} *)
+
+type (_, _) eq = Refl: ('a, 'a) eq
```
Comment (xavierleroy) [inline stdlib/stdlib.mli]: For the record: I'm not a fan of the `Refl` name, which is jargon, just like "lambda" is jargon.

File: stdlib/type.mli [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(** Type introspection.
+
+    @since 5.1 *)
+
+(** {1 Type equality witness} *)
```
Comment (dbuenzli) [inline stdlib/type.mli]: Could you please add an anchor here, e.g. `{1:witness }`.

**Review by dbuenzli — APPROVED** [review]
Regarding naming I made my preference known [here](https://github.com/ocaml/ocaml/issues/11564#issuecomment-1270020417) but I'm fine with this.

File: stdlib/StdlibModules [hunk 3]
```diff
@@ -44,7 +44,7 @@ STDLIB_MODULE_BASENAMES = \
   fun gc digest bigarray random hashtbl weak \
   format scanf callback camlinternalOO oo camlinternalMod ephemeron \
   filename complex arrayLabels listLabels bytesLabels stringLabels moreLabels \
-  stdLabels in_channel out_channel effect
+  stdLabels in_channel out_channel effect type
```
Comment (dbuenzli) [inline stdlib/StdlibModules]: Just a nit since there's not much to say :-) `STDLIB_MODULE_BASENAME` is supposed to be in dependency order, I think it would be better to have `type` further up so that the day people want to use it somewhere else in the stdlib they don't get hampered.

Comment (gasche) [reply stdlib/StdlibModules]: (If you have the mental fortitude to do it, don't hesitate to also add line breaks to this definition so that there is only one item per line. It's annoying when you do it first, but then generations of committers and reviewers are going to thank you for the absence of paragraph-filling worries and the clean diffs.)

File: stdlib/type.mli (hunk 2 above)
Comment (nojb) [reply stdlib/type.mli]: Fixed, thanks.

File: stdlib/StdlibModules (hunk 3 above)
Comment (nojb) [reply stdlib/StdlibModules]: Done (for both points), thanks.

File: stdlib/StdlibModules [hunk 4]
```diff
@@ -34,17 +34,72 @@
...
-  seq option result bool char uchar \
-  list int bytes string unit marshal array float int32 int64 nativeint \
-  lexing parsing set map stack queue buffer \
-  mutex condition semaphore domain \
-  camlinternalFormat printf arg printexc \
-  fun gc digest bigarray random hashtbl weak \
-  format scanf callback camlinternalOO oo camlinternalMod ephemeron \
-  filename complex arrayLabels listLabels bytesLabels stringLabels moreLabels \
-  stdLabels in_channel out_channel effect
+  type \
```
Comment (dbuenzli) [inline stdlib/StdlibModules]: Is it ok to put type before `stdlib` ?

Comment (dbuenzli) [reply stdlib/StdlibModules]: scratch that :-) rather is it ok to have stdlib that further up.

Comment (nojb) [reply stdlib/StdlibModules]: Well, perhaps it works because `Type` currently does not depend on anything at all, but the convention is that we put all the "wrapped" modules after `stdlib` so I moved it.

> scratch that :-) rather is it ok to have stdlib that further up.

This is always a bit confusing, but I think that the `-no-alias-deps` stuff means that `stdlib` does not actually depend on any other module (while in turn every other module depends on it to resolve the "wrapped" names).

**Comment by garrigue** [comment]
For the record, putting such a commonly used type definition inside a submodule still strikes me as against the current design of the standard library.
I understand putting types together with the operations that work on them in the same module, but here `eq` can be used on its own, just like `result` or `option`.

And I don't buy the argument of namespace pollution: you can perfectly use the same names in your own programs without conflict. Actually, it goes the other way: with the current PR, there is only one definition inside `Type`, so I can just do `open Type` to use the definition comfortably, but if we start adding functions to `Type`, then it would import them, even if I don't need them.

By the way, I'm perfectly fine with calling the constructor `Equal`.

**Comment by dbuenzli** [comment]
> I understand putting types together with the operations that work on them in the same module, but here `eq` can be used on its own, just like `result` or `option`.

See [this comment](https://github.com/ocaml/ocaml/issues/11564#issuecomment-1265220734) by @alainfrisch but I agree it would be nice to have a type for equality functions. Something like:

```
type 'a Stdlib.eq = 'a -> 'a -> bool
```

> so I can just do `open Type` to use the definition comfortably, but if we start adding functions to `Type`, then it would import them, even if I don't need them.

Your readers will likely be comforted to read `Type.Equal`.

**Comment by garrigue** [comment]
I'm sorry, but I don't see any kind of ambiguity here. If you look at the definition in `Stdlib`, you will immediately see what it is about.  We are not talking about an abstract type here.
My point is that, while some types are re-exported by submodules, all concrete types are defined inside `Stdlib` itself, except for those that are _only_ used in combination with the operation in that submodule. So we would be departing from this design principle.

**Comment by dbuenzli** [comment]
> I'm sorry, but I don't see any kind of ambiguity here. If you look at the definition in `Stdlib`, you will immediately see what it is about.

Good names are precisely made to avoid trips to definitions… If you want we could have: 

```
type (_, _) Stdlib.type_eq = Type_equal : ('a, 'a) type_eq
```

But I think it would be rather silly given that there are already plans to add things to `Type`. With the `Type` module we have a nice namespace to organize that.

Other than that, in general, when you deal with large practical programs it is better to leave unqualified names to user space (as in definitions of the current module). It makes it easier to orient. A large unqualified scope is a cognitive burden and increases the probability of misreads (because the probability of redefinitions increases).

> My point is that, while some types are re-exported by submodules, all concrete types are defined inside `Stdlib` itself, except for those that are _only_ used in combination with the operation in that submodule. So we would be departing from this design principle.

I wasn't aware this was a design principle. But design principle are made to be questioned – and not followed if they can't justify themselves. So:

1. What good things are provided to end-users by following this design principle here ? 
2. What are the drawbacks for end-users if this design principle is not followed here ?

**Comment by gasche** [comment]
Maybe a compromise would be to have the type definition in `stdlib`, and re-export it in a new `Type` module of the stdlib with no new definition? We would both satisfy @garrigue's design constraint and the others' naming preference and have a proper "place to add more features". (Downside: it's a bit silly to have a module with only one type definition in it. But I don't really see a problem with that.)

Naming: in this approach both definitions must share the same constructor name, not `Type_equal` in Stdlib and `Equal` in Type, but the type name can be more explicit in Stdlib to compensate for the non-qualification.

```ocaml
(* stdlib.mli *)
type (_, _) type_equality = Equal : ('a, 'a) type_equality

(* type.mli *)
type ('a, 'b) eq = ('a, 'b) Stdlib.type_equality = Equal : ('c, 'c) eq
```

**Comment by dbuenzli** [comment]
> Maybe a compromise would be to have the type definition in stdlib, and re-export it in a new Type module of the stdlib with no new definition? 

The more definitions and (even inconsistent in your proposal) names the better… That's a given in the OCaml eco-system :–) 

> We would both satisfy @garrigue's design constraint

Before starting to compromise, why don't we try to have reasonable answers to the simple questions I asked about that design constraint ?

**Comment by garrigue** [comment]
If you really think that we are talking about a seldom used type such that people need a long name to understand it, then put it in `Type` and be done with it. At least one can use `open Type` to use it.

But if we are talking about a standard type used all the time with GADTs, like `option` or `list`, then it makes sense to keep a short name and have it in `Stdlib`.

As for the design principles of the standard library, I don't think they are stated anywhere explicitly, I just observe how it is organized, and state an observation. My fuzzy feeling is that changing this may lead to confusion, but I may be wrong.
As to why it is organized that way, I think this is related to how are organized many smallish programs, where one puts all the types in one common definition module, that is always open, to avoid having to open other modules. Whether this is a good design or not is open to discussion, in particular I agree this is not the best way to create a large library (the standard library is relatively small), but this happens to be the way it was built. (In older versions of ML, many modules were open by default, creating an even flatter namespace.)

**Comment by dbuenzli** [comment]
> If you really think that we are talking about a seldom used type such that people need a long name to understand it, then put it in `Type` and be done with it. At least one can use `open Type` to use it.

Or use `Type.Equal` which doesn't feel egregious and has the advantage of clarity.

> But if we are talking about a standard type used all the time with GADTs, like `option` or `list`, then it makes sense to keep a short name and have it in `Stdlib`.

Sure it's nice to have short notations for `option`, `result` and `list`. I'm less convinced by the need for this type though. 

My point is less about the problem of having it in the `Stdlib` or a dedicated module than having a good descriptive name for it. An unqualified `Equal` variant could be used for many other things in user space. I count at least two `` `Equal``s and one `Equals` in my code base which are unrelated to the `Equal` proposed here (for which I have been using `Eq`). 

Using `Type_equal`  makes it obvious what you are talking about (and at which level you are playing). But again, since there are plans for a `Type` module, that screams for `Type.Equal`, same number of characters, reads well, is properly namespaced and you can still use it unqualified if that's your thing.

> As for the design principles of the standard library, I don't think they are stated anywhere explicitly, I just observe how it is organized, and state an observation.

So I gather that it's rather an observation than a design principle. The observation seems partial though, there are already other types with concrete representations that are solely defined in modules, a quick look suggests at least `Complex`, `Either` and `Seq`.

**Comment by Octachron** [comment]
One small remark I think that for the `(_,_) eq` type there is an important difference between the visibility need of the type and the constructor.

The type `(_,_) eq` needs to appear in type annotations to be useful, and it is mostly the main type that may appear in this context with this name. By contrast, the `Equal` constructor of this type have a less clear meaning on its own.

I wouldn't mind having to write:

```ocaml
let transitivity (type a b c d) (x: (a,b) eq) (y: (b,c) eq) (z: (c,d) eq) : (a,d) eq =
  match x, y, z with
  | Type.Equal, Type.Equal, Type.Equal -> Type.Equal
```
(Even if type-directed disambiguation allow to elide the `Type.` prefix).
However, 
```ocaml
let transitivity (type a b c d) (x: (a,b) Type.eq) (y: (b,c) Type.eq) (z: (c,d) Type.eq) : (a,d) Type.eq =
  match x, y, z with
  | Type.Equal, Type.Equal, Type.Equal -> Type.Equal
```
is harder to read.
And far as I see, this is the reason why @garrigue  is focusing on the type name whereas @dbuenzli is more keen on discussing the constructor name.

Would it make sense to have:
```ocaml
(* stdlib.mli *)
type ('a,'b) eq =  ('a,'b) Type.eq
```
and
```ocaml
(* type.mli *)
type ('a,'b) eq = Equal: ('a,'b) eq
```
with the idea that the name of the type should be as short as possible whereas it is better for the constructor to have a self-explaining full name?

**Comment by dbuenzli** [comment]
> The type (_,_) eq needs to appear in type annotations to be useful,

That's not my own experience. As far as I'm concerned I don't think I have ever written a `Type.eq` type except in the module where type identifiers are defined (and not more than once in a sig, which feels manageable). 

All my uses are indirect via type ids in which you do not write this type to use it. For example:

```ocaml
let find : type a. a key -> t -> a option =
  fun k m -> match M.find_opt k.Key.untyped m with
  | None -> None
  | Some B (k', v) ->
      match Tid.equal k.Key.tid k'.Key.tid with
      | None -> None | Some Tid.Eq -> Some v
```


So the question is, how widespread do you think code like you mention is ? 

Is it widespread enough to consider an addition to `Stdlib` ? Isn't it sufficient to simply `open Type` or even define your own alias locally for these cases ?

I really think the `Stdlib` namespace should be treated with economy and additions be exceptional.

**Review by garrigue — APPROVED** [review]
I think this is a good starting point.
Discussed it with @nojb and @Octachron

**Comment by nojb** [comment]
> I think this is a good starting point. Discussed it with @nojb and @Octachron

Thanks @garrigue! I rebased the PR on `trunk`. I believe the current version satisfies (to different degrees) everyone who has expressed an opinion. On that note, I am planning to merge by end of day, today. Yell if you have something to add!

**Comment by nojb** [comment]
(I also pushed a revised, slightly expanded, docstring.)

File: manual/src/library/stdlib-blurb.etex [hunk 5]
```diff
@@ -177,6 +178,7 @@ be called from C \\
 \stddocitem{String}{string operations}
 \stddocitem{StringLabels}{string operations (with labels)}
 \stddocitem{Sys}{system interface}
+\stddocitem{Type}{type instrospection}
```
Comment (yallop) [inline manual/src/library/stdlib-blurb.etex]:
```suggestion
\stddocitem{Type}{type introspection}
```

File: stdlib/type.mli [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+(** Type introspection.
+
+    @since 5.1 *)
+
+(** {1:witness Type equality witness} *)
+
+type (_, _) eq = Equal: ('a, 'a) eq
+(** The purpose of [eq] is to represent type equalities that may not otherwise
+    be known by the type checker (eg because they may depend on dynamic data).
```
Comment (yallop) [inline stdlib/type.mli]:
```suggestion
    be known by the type checker (e.g. because they may depend on dynamic data).
```

File: stdlib/type.mli [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    equal.
+
+    If one has a value [eq : (a, b) eq] that proves types [a] and [b] are equal,
+    one can use it to convert a value of type [a] to a value of type [b] by
+    pattern matching on [Equal]:
+    {[
+      let cast (type a) (type b) (Equal : (a, b) Type.eq) (a : a) : b = a
+    ]}
+
+    At runtime, this function is just the identity.
```
Comment (yallop) [inline stdlib/type.mli]: Not quite: it has two parameters

File: manual/src/library/stdlib-blurb.etex (hunk 5 above)
Comment (nojb) [reply manual/src/library/stdlib-blurb.etex]: Thanks, fixed.

File: stdlib/type.mli (hunk 6 above)
Comment (nojb) [reply stdlib/type.mli]: Thanks, fixed.

File: stdlib/type.mli (hunk 7 above)
Comment (nojb) [reply stdlib/type.mli]: Good point. I changed it to `At runtime, this function acts as the identity on [a]`. What do you think?

Comment (yallop) [reply stdlib/type.mli]: I think it's better, but I'd just say something like "At runtime, this function simply returns its second argument unchanged"

Comment (nojb) [reply stdlib/type.mli]: Sounds good to me. Amended as suggested. Thanks!
